### ISSUE TITLE
Case where the parallel_backend doesn't propagate through nested Parallel calls

### ISSUE THREAD

**Issue by TomAugspurger** [joblib/joblib#615]
This requires distributed master to run.

```python
from sklearn.externals import joblib
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score

from distributed import Client
import distributed.joblib

client = Client()


X, y = make_classification(n_samples=10_000)

address = client.scheduler_info()['address']

param_grid = {'min_samples_split': [2, 5]}
cv = GridSearchCV(RandomForestClassifier(), param_grid, cv=4, n_jobs=-1)

with joblib.parallel_backend('dask', scheduler_host=address,
                             scatter=[X, y]):
    cross_val_score(cv, X, y, cv=2, n_jobs=1, verbose=50)
```

I observe that the LokyBackend is being created (and so I don't get good nested parallelism).

In the outer call to `Parallel`, in `cross_val_score` I think, things seem fine.
In the nested calls to `Parallel()`, I see in https://github.com/joblib/joblib/blob/f8384e182a119c786ba21693776d683434107c70/joblib/parallel.py#L55-L59 that `active_backend_and_jobs` is None, and so the default backend (Loky) is being used. Still debugging.

**Comment by TomAugspurger** [comment]
Potentially simpler example, though not sure if it's the same cause.

```python
from distributed import Client
import distributed.joblib
import joblib

def f(dask_worker):
    from sklearn.externals.joblib.parallel import get_active_backend

    return get_active_backend(verbose=10)

client = Client()
address = client.scheduler_info()['address']
with joblib.parallel_backend('dask', scheduler_host=address):
    res = client.run(f)
res
```

```
{'tcp://127.0.0.1:64570': (<sklearn.externals.joblib._parallel_backends.LokyBackend at 0x10f833400>,
  1),
 'tcp://127.0.0.1:64572': (<sklearn.externals.joblib._parallel_backends.LokyBackend at 0x10f8333c8>,
  1),
 'tcp://127.0.0.1:64573': (<sklearn.externals.joblib._parallel_backends.LokyBackend at 0x10f833390>,
  1),
 'tcp://127.0.0.1:64576': (<sklearn.externals.joblib._parallel_backends.LokyBackend at 0x10f833978>,
  1),
 'tcp://127.0.0.1:64578': (<sklearn.externals.joblib._parallel_backends.LokyBackend at 0x10f833518>,
  1),
 'tcp://127.0.0.1:64579': (<sklearn.externals.joblib._parallel_backends.LokyBackend at 0x10f8332b0>,
  1),
 'tcp://127.0.0.1:64581': (<sklearn.externals.joblib._parallel_backends.LokyBackend at 0x10f833908>,
  1),
 'tcp://127.0.0.1:64598': (<sklearn.externals.joblib._parallel_backends.LokyBackend at 0x10f833940>,
  1)}
```

Should we expect the values to be `DaskDistributedBackend`s?

**Comment by TomAugspurger** [comment]
Specifically, since the `_backend` is a thread local storing `backend_and_jobs`, should we expect it to pick up the backend from the outer context manager?

**Comment by TomAugspurger** [comment]
Perhaps a clearer example than https://github.com/joblib/joblib/issues/615#issuecomment-360781478

```python
from distributed import Client
import distributed.joblib
import joblib

def f(dask_worker):
    p = joblib.Parallel(n_jobs=2)
    return p._backend

client = Client()
address = client.scheduler_info()['address']
with joblib.parallel_backend('dask', scheduler_host=address) as (b, nj):
    res = client.run(f)

res
```

output

```python
{'tcp://127.0.0.1:49248': <joblib._parallel_backends.LokyBackend at 0x1159a8fd0>,
 'tcp://127.0.0.1:49250': <joblib._parallel_backends.LokyBackend at 0x1159a81d0>,
 'tcp://127.0.0.1:49255': <joblib._parallel_backends.LokyBackend at 0x1159a8198>,
 'tcp://127.0.0.1:49259': <joblib._parallel_backends.LokyBackend at 0x1159a85f8>,
 'tcp://127.0.0.1:49261': <joblib._parallel_backends.LokyBackend at 0x1159a8be0>,
 'tcp://127.0.0.1:49262': <joblib._parallel_backends.LokyBackend at 0x1159a8f98>,
 'tcp://127.0.0.1:49265': <joblib._parallel_backends.LokyBackend at 0x1159a8828>,
 'tcp://127.0.0.1:49266': <joblib._parallel_backends.LokyBackend at 0x1159a8160>}
```

Whereas, from the "main thread", things are correct

```python
with joblib.parallel_backend('dask', scheduler_host=address) as (b, nj):
    p = joblib.Parallel(n_jobs=2)
p._backend
```
<distributed.joblib.DaskDistributedBackend at 0x116368390>

**Comment by TomAugspurger** [comment]
So, perhaps this is what's happening:

In the

```
with joblib.parallel_backend("dask", schedeler_host=address):
    p = joblib.Parallel(n_jobs=2)
```

The `Parallel.__init__` is called on a thread whose `_backend` thread local has been set to `'dask'`.

But the `with joblib.parallel_backend` context manager does *not* affect the `_backend` thread local of the dask workers, because they're in separate threads / processes. It seems like we need some other way of communicating the "active backend, as set by the user" than a thread local. I have no suggestions on how to do that yet :)

**Comment by ogrisel** [comment]
When calling with `client.run` as done in https://github.com/joblib/joblib/issues/615#issuecomment-360784207 it's normal that we loose the joblib backend info. But it should not be the case if we use a nested call to `joblib.Parallel` instead of `client.run`.

**Comment by TomAugspurger** [comment]
tldr: user error and @ogrisel's code is perfect.

`cross_val_score` wasn't getting perfect parallelism with distributed, since it's calling `RandomForestClassifier.predict`, which *requires* shared memory, so joblib is doing the correct thing.

**Comment by ogrisel** [comment]
> @ogrisel's code is perfect

It's not so easy to follow though ;)

Let's close!
